### ISSUE TITLE
Many to many entity add faild

### ISSUE THREAD

**Issue by smith-kevin** [dotnet/efcore#29882]
## File a bug

I have a many to many relationship entity. When I add a proxy surrogate to the relationship, the first time I add an entity to the database will succeed, but the second time I add an entity, an exception of KF conflict will be thrown.When there is no surrogate key in the relationship, the program will not exception.This Issue is similar to #24466.

### Include your code

```C#
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

using Context context = new Context();
await context.Database.EnsureDeletedAsync();
await context.Database.EnsureCreatedAsync();

var tag = new Tag();
await context.AddAsync(tag);
await context.SaveChangesAsync();

var blog1 = new Blog() { Name = "test1", Tags = new List<Tag> { tag } };
var blog2 = new Blog() { Name = "test2", Tags = new List<Tag> { tag } };

//This command can be executed successfully.
await context.AddAsync(blog1);
await context.SaveChangesAsync();

//An error will be reported when adding the second time.When I delete the surrogate key, no exception will be thrown.
//The INSERT statement conflicted with the FOREIGN KEY constraint "FK_BlogTag_Blog_BlogId". The conflict occurred in database "Test", table "dbo.Blog", column 'Id'.
await context.AddAsync(blog2);
await context.SaveChangesAsync();

Console.ReadLine();


public class Blog
{
	[Key]
	public int Id { get; set; }
	public string Name { get; set; }

	public List<Tag> Tags { get; set; }
	public List<BlogTag> BlogTags { get; set; }
}

public class Tag
{
	[Key]
	public int Id { get; set; }

	public List<Blog> Blogs { get; set; }
	public List<BlogTag> BlogTags { get; set; }
}

public class BlogTag
{
	//When there is no surrogate key in the relationship, the program will not exception.
	[Key]
	public int Id { get; set; }

	public int BlogId { get; set; }
	public int TagId { get; set; }

	[ForeignKey(nameof(BlogId))]
	public Blog Blog { get; set; }

	[ForeignKey(nameof(TagId))]
	public Tag Tag { get; set; }
}

public class Context : DbContext
{

	protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
	{
		base.OnConfiguring(optionsBuilder);
		optionsBuilder.UseSqlServer("Server=.;Database=Test;User Id=sa;Password=123456;Trust Server Certificate=true;")
			.EnableSensitiveDataLogging()
			.LogTo(Console.WriteLine,LogLevel.Information);
	}

	protected override void OnModelCreating(ModelBuilder modelBuilder)
	{
		modelBuilder.Entity<Blog>()
			.HasMany(t => t.Tags)
			.WithMany(t => t.Blogs)
			.UsingEntity<BlogTag>(
				j => j.HasOne(t => t.Tag).WithMany(t => t.BlogTags).HasForeignKey(t => t.TagId),
				j => j.HasOne(t => t.Blog).WithMany(t => t.BlogTags).HasForeignKey(t => t.BlogId)
			);

		modelBuilder.Entity<Tag>();
		modelBuilder.Entity<BlogTag>();
	}
}

```

### Include stack traces

```
fail: 2022/12/17 21:13:23.642 CoreEventId.SaveChangesFailed[10000] (Microsoft.EntityFrameworkCore.Update)
      An exception occurred in the database while saving changes for context type 'Context'.
      Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while saving the entity changes. See the inner exception for details.
       ---> Microsoft.Data.SqlClient.SqlException (0x80131904): INSERT 语句与 FOREIGN KEY 约束"FK_BlogTag_Blog_BlogId"冲突。该冲突发生于数据库"demo"，表"dbo.Blog", column 'Id'。
      语句已终止。
         at Microsoft.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__188_0(Task`1 result)
         at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
         at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
         at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
      --- End of stack trace from previous location ---
         at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
         at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
      --- End of stack trace from previous location ---
         at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
      ClientConnectionId:22cdc696-7f99-41a3-96c1-fc3e1aa51c75
      Error Number:547,State:0,Class:16
         --- End of inner exception stack trace ---
         at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(IEnumerable`1 commandBatches, IRelationalConnection connection, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IList`1 entriesToSave, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(StateManager stateManager, Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
```

### Log when added for the first time

```
info: 2022/12/17 21:13:23.534 RelationalEventId.CommandExecuted[20101] (Microsoft.EntityFrameworkCore.Database.Command)
      Executed DbCommand (54ms) [Parameters=[@p0='test1' (Nullable = false) (Size = 4000)], CommandType='Text', CommandTimeout='30']
      SET NOCOUNT ON;
      INSERT INTO [Blog] ([Name])
      VALUES (@p0);
      SELECT [Id]
      FROM [Blog]
      WHERE @@ROWCOUNT = 1 AND [Id] = scope_identity();
info: 2022/12/17 21:13:23.551 RelationalEventId.CommandExecuted[20101] (Microsoft.EntityFrameworkCore.Database.Command)
      Executed DbCommand (4ms) [Parameters=[@p1='1', @p2='1'], CommandType='Text', CommandTimeout='30']
      SET NOCOUNT ON;
      INSERT INTO [BlogTag] ([BlogId], [TagId])
      VALUES (@p1, @p2);
      SELECT [Id]
      FROM [BlogTag]
      WHERE @@ROWCOUNT = 1 AND [Id] = scope_identity();
```

### Log when added for the second time
```
info: 2022/12/17 21:13:23.558 RelationalEventId.CommandExecuted[20101] (Microsoft.EntityFrameworkCore.Database.Command)
      Executed DbCommand (2ms) [Parameters=[@p0='test2' (Nullable = false) (Size = 4000)], CommandType='Text', CommandTimeout='30']
      SET NOCOUNT ON;
      INSERT INTO [Blog] ([Name])
      VALUES (@p0);
      SELECT [Id]
      FROM [Blog]
      WHERE @@ROWCOUNT = 1 AND [Id] = scope_identity();
fail: 2022/12/17 21:13:23.573 RelationalEventId.CommandError[20102] (Microsoft.EntityFrameworkCore.Database.Command)
      Failed executing DbCommand (6ms) [Parameters=[@p1='-2147482646', @p2='1'], CommandType='Text', CommandTimeout='30']
      SET NOCOUNT ON;
      INSERT INTO [BlogTag] ([BlogId], [TagId])
      VALUES (@p1, @p2);
      SELECT [Id]
      FROM [BlogTag]
      WHERE @@ROWCOUNT = 1 AND [Id] = scope_identity();
```

### Include provider and version information

EF Core version: 6.0.12
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 6.0
Operating system: Windows 10
IDE: Visual Studio 2022 17.4.2

**Comment by ajcvickers** [comment]
Note for triage: repros on 7.0. The value generated for `Blog.Id` in the first insert is not being propagated to the `BlogTag.BlogId` in the second insert, resulting in an attempt to insert a temporary value instead.

**Comment by ajcvickers** [comment]
Workaround is to insert `blog2` first, then add the tag and save again:

```C#
var blog1 = new Blog() { Name = "test1", Tags = new List<Tag> { tag } };
var blog2 = new Blog() { Name = "test2", Tags = new List<Tag>() };
await context.AddAsync(blog1);
await context.SaveChangesAsync();

await context.AddAsync(blog2);
await context.SaveChangesAsync();

blog2.Tags.Add(tag);
await context.SaveChangesAsync();
```

**Comment by AndriySvyryd** [comment]
Fixed in e1d0a59a90cd8f9cc883f9a3ad4ecdbe9a9c833e
